source/extensions/transport_sockets/tls/io_handle_bio.cc

```diff
@@ -0,0 +1,140 @@
+#include "extensions/transport_sockets/tls/io_handle_bio.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/network/io_handle.h"
+
+#include "openssl/bio.h"
+#include "openssl/err.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+
+namespace {
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+inline Envoy::Network::IoHandle* bio_io_handle(BIO* bio) {
+  return reinterpret_cast<Envoy::Network::IoHandle*>(bio->ptr);
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+int io_handle_new(BIO* bio) {
+  bio->init = 0;
+  bio->num = -1;
+  bio->ptr = nullptr;
+  bio->flags = 0;
+  return 1;
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+int io_handle_free(BIO* bio) {
+  if (bio == nullptr) {
+    return 0;
+  }
+
+  if (bio->shutdown) {
+    if (bio->init) {
+      bio_io_handle(bio)->close();
+    }
+    bio->init = 0;
+    bio->flags = 0;
+  }
+  return 1;
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+int io_handle_read(BIO* b, char* out, int outl) {
+  if (out == nullptr) {
+    return 0;
+  }
+
+  Envoy::Buffer::RawSlice slice;
+  slice.mem_ = out;
+  slice.len_ = outl;
+  auto result = bio_io_handle(b)->readv(outl, &slice, 1);
+  BIO_clear_retry_flags(b);
+  if (!result.ok()) {
+    if (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again) {
+      BIO_set_retry_read(b);
+    }
+    return -1;
+  }
+  return result.rc_;
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+int io_handle_write(BIO* b, const char* in, int inl) {
+  Envoy::Buffer::RawSlice slice;
+  slice.mem_ = const_cast<char*>(in);
+  slice.len_ = inl;
+  auto result = bio_io_handle(b)->writev(&slice, 1);
+  BIO_clear_retry_flags(b);
+  if (!result.ok()) {
+    if (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again) {
+      BIO_set_retry_write(b);
+    }
+    return -1;
+  }
+  return result.rc_;
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+long io_handle_ctrl(BIO* b, int cmd, long num, void* ptr) {
+  long ret = 1;
+
+  switch (cmd) {
+  case BIO_C_SET_FD:
+    io_handle_free(b);
+    b->num = -1;
+    b->ptr = ptr;
+    b->shutdown = int(num);
+    b->init = 1;
+    break;
+  case BIO_C_GET_FD:
+    ret = -1;
+    *reinterpret_cast<void**>(ptr) = b->ptr;
+    break;
```

Comment:
Do we really need to implement BIO_C_SET_FD and BIO_C_GET_FD?

It may be good to add ASSERT(false) for cases that we do not expect to ever happen at the proxy.  This includes: BIO_CTRL_RESET, BIO_C_FILE_SEEK, BIO_C_FILE_TELL, BIO_CTRL_INFO, BIO_C_GET_FD, BIO_C_SET_FD:

Reply:
I've used the socket bio as an example (since we're faking one) and those are the only control messages it implements. BIO_C_SET_FD is the one I've used for setting the ptr but I guess we could change that. 
What would you prefer, keep following the socket model or implement things differently?

Reply:
Your constructor could set b->ptr directly instead of using BIO_C_SET_FD.

I think  your implementation of iohandle BIOs is a great improvement, I'm just working my way through some nits.

Reply:
Ow. Yes we can, but we need to initialize all the other fields (shutdown, init, num). Should I do that explicitly in the constructor and then ASSERT(false) on BIO_C_SET_FD and BIO_C_GET_FD?

Reply:
Yes, we'ld need to also initiate other relevant fields like init also in the contructor.

Reply:
Perfect! Done!